### ISSUE TITLE
Iterator protocol for the listing APIs

### ISSUE THREAD

**Issue by isidentical** [intake/filesystem_spec#632]
Currently the listing APIs only return the listings when everything is read (e.g `ls()` / `find()` / `glob()`) though this is blocking applications which normally would be able to process the information as it comes. One example is the DVC, where we display progress bars to make the user aware that stuff is happening in the background and estimate when it is going to be potentially over. 

Even though there are multiple listing APIs (including glob/walk), I think the starter would be the ones that are mostly used (find/ls). The default implementation will fall back to this form;
```py
def lazy_find(self, path, *args, **kwargs):
    results = self.find(path, *args, **kwargs):
    yield from results
```

though the actual forms will be implemented in each cloud (s3, gs, azure, ssh, etc) and hopefully they will yield the results as the pagination gives them so that the users would be able to see the progress between different API calls that might potentially take seconds. 

One important thing is that async to sync infrastruct `(sync_wrapper)` is not suitable for the async iterators since everything assumes the stuff is `await`able but we need to actually wrap those with `async for ...` and yield each item. Other than that, I don't see major pitfalls that we might anticipate on the fsspec itself though it might be a little bit weird for the implementations to dupllicate the listing code, so one other consideration would be make the original implementation lazy by default (for example make `_find` `_lazy_find`) and then implement `_find` by just `[file async for file in self._lazy_find(path)]`, though this might be something for the future where implementations themselves can decide).

**Comment by martindurant** [comment]
I am not opposed in principle, but I do worry how much it might complicate the code.

Note that listing functions typically are paginated internally (as you say), so if the specific use case is of updating the UI, then this sounds more like a callbacks mechanism than iterator protocol. On the other hand, it may be useful to have the iterator approach more simply to be able halt pagination mid-way rather than accumulate a massive list every time.

Of course, the whole idea mixes in an awkward way with directory caching. I suspect it would be a completely alternate path.

**Comment by isidentical** [comment]
> then this sounds more like a callbacks mechanism than iterator protocol.

Progress bars and a decent callback system is nice, though that is not the only issue that can be solved by this. Another simple example would be starting the real process with chunks of the tree before waiting `.find()` to traverse millions of files. That is the nice thing about an iterator protocol, it removes all the restrictions and the blocking aspect of the code.

> On the other hand, it may be useful to have the iterator approach more simply to be able halt pagination mid-way rather than accumulate a massive list every time.

Indeed

> Of course, the whole idea mixes in an awkward way with directory caching. I suspect it would be a completely alternate path.

If it get completes, then there is nothing that prevents fsspec to cache to contents. If it halts on the halfway, then there won't be any issue regarding the caching, since the code path will never get to be executed. But yea, initially we could have 2 parallel implementations and then use the lazy API for the blocking API too (since when you are just doing `list(self.lazy_find())` the behavior should be the exactly same version of the normal `self.find()` implementation.

**Comment by isidentical** [comment]
Another example within DVC is determining whether a directory is empty or not. So instead of traversing all the files, we could just check whether there is any files in the first iteration and quit if so. This would mean that we won't be caching the results, which I believe is the desired behaviour for this case. 

https://github.com/iterative/dvc/blob/c6f49bd23d172a9831fc4b2e0515ff3f59ca7787/dvc/fs/fsspec_wrapper.py#L84-L85

**Comment by martindurant** [comment]
We already do MaxKeys=1 for the purpose of exists/info/isdir on a directory :)

**Comment by isidentical** [comment]
Yeah, makes sense (and will probably use it). Though just mentioned it as a likely candidate to the actual proposal.

**Comment by lgeiger** [comment]
This would also be useful for `cat()` when it is called with either multiple paths or with a glob pattern. That way not only a progressbar can be integrated nicely, but there is also no need to keep all the content in memory which might be problematic when reading large amounts of data.

**Comment by martindurant** [comment]
Note that implementing the callback mechanism (already done for get/put) would be a simpler way to implement a progressbar.

Having said that, I am not against the idea, it's more of a matter of how to implement it and fit it into the API.

**Comment by lgeiger** [comment]
> it's more of a matter of how to implement it and fit it into the API.

I agree, the API is already quite complicated since many functions might either return a single value or a dictionary including multiple ones. For me intuitively it would actually make the most sense that everything would lazily yield instead of returning a dictionary with multiple values for such cases while keeping the current behaviour whenever the return value is a single one. However, this would be a pretty breaking change (similar to Python 2 map vs Python 3 map), so not sure if that's an option.

I guess the least disruptive change would be adding a boolean keyword argument like `lazy`, `return_dict`, `yield_values`, `return_iterator`, etc which defaults to the current behaviour which can be progressively adopted instead of duplicating the API surface by adding lazy versions of the existing API. If something like this would be welcome, I'm happy to try and prototype a solution to see how this would fit in with the current async to sync infrastructure.
